### ISSUE TITLE
vax_sys.c: GETNUM and PUTNUM macros

### ISSUE THREAD

**Issue by bscottm** [simh/simh#772]
vax_sys.c might have an interesting array bound/memory overwrite issue in the `GETNUM` and `PUTNUM` macros:

```c
#define GETNUM(d,n)     for (k = d = 0; k < n; k++) \
                    d = d | (((int32) val[vp++]) << (k * 8))

#define PUTNUM(d,n)     for (k = 0; k < n; k++) val[vp++] = (d >> (k * 8)) & 0xFF
```

`vp` gets incremented by `k` each time these macros are called. But if you look at the references to `parse_sym` in the `scp.c` driver, the `val` parameter is usually a pointer to a single `t_value`. Consequently, it looks like there's an array bound/memory overwrite issue.

@archivisth: Concur?

The fix is probably an array local to `parse_sym` that accumulates for `PUTNUM` and calculates the value for `GETNUM` and assigns to `val`.

(Note: This reduces some pointer wrapping warnings when compiling with `-fstrict-overflow`.)

**Comment by markpizz** [comment]
The code paths which invoke parse_sym are several and complicated with only one actually providing a pointer to a single member t_value.  In this specific case, the flag SIM_SW_REG is set, indicating that we're parsing a register value, and then parse_sym is called ONLY if one of the REG_VMFLAGS flags are set in the REGister definition.  None of the VAX CPU registers have these flag bits set, so in the register case we never traverse the parse_sym code path for this simulator.

**Comment by bscottm** [comment]
@mark: `#if 0` the code out and always return `SCPE_OK`?

**Comment by archivisth** [comment]
What Mark said ;)

The VAX use case for fprint_sym and parse_sym is to print the contents of the string of bytes in array sim_eval or to parse a character string and place the resulting bytes in sim_eval. The macros are designed for that.

REG_VMIO override was added for simulators that have fixed length words. The only place it is used in my simulators is the LGP-30, allowing registers that hold addresses to be displayed/parsed in the LGP-30's peculiar address encoding scheme.

You can't #if 0 code that is used in other simulators. The VAX is just one of of dozens of simulators in SimH.

**Comment by markpizz** [comment]
> @markpizz: #if 0 the code out and always return SCPE_OK?

No, I don't think so.  DEVICES (other than the CPU) may exist now or be added later which depend on this functionality for register behavior, and parse_sym  is used in many places so parse_sym can't be changed to #if 0

Some warnings produced by static code analysis are false positives.  It would seem that this is one of them.

**Comment by bscottm** [comment]
It wasn't something that `cppcheck` found. I noticed it while browsing the code to isolate where GCC complains about potential signed overflow that it optimized.

BTW: My suggestion was to `#if 0` the code in `vax_sys.c`, and nowhere else, since that code is effectively dead code (never called.) Maybe even call `sim_messagef` to output a message in the unlikely case that VAX's `parse_sym` ever gets called.

The particular "danger", such as it is, is the use of function arguments for something other than what they're intended. In this case, `val` is assumed to be an array with more than one element when the call sites don't pass an array with more than one element.

**Comment by bscottm** [comment]
(FWIW: My first computing experience was depositing the PDP8 RIM loader, followed by Trash-80's, Apples in high school and VAX/750s on a HSC cluster in college.)
